crates/valence_generated/build/effect.rs

```diff
@@ -0,0 +1,218 @@
+                EffectCategory::Harmful => quote! { EffectCategory::Harmful },
+                EffectCategory::Neutral => quote! { EffectCategory::Neutral },
+            };
+
+            let name = ident(effect.name.to_pascal_case());
+            quote! {
+                Self::#name => #category,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let effect_to_color_arms = effects
+        .iter()
+        .map(|effect| {
+            let color = &effect.color;
+            let name = ident(effect.name.to_pascal_case());
+            quote! {
+                Self::#name => #color,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let effect_to_instant_arms = effects
+        .iter()
+        .map(|effect| {
+            let instant = &effect.instant;
+            let name = ident(effect.name.to_pascal_case());
+            quote! {
+                Self::#name => #instant,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let effect_variants = effects
+        .iter()
+        .map(|effect| ident(effect.name.to_pascal_case()))
+        .collect::<Vec<_>>();
+
+    Ok(quote! {
+        use valence_ident::{Ident, ident};
+
+        /// Represents an effect category
+        #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+        pub enum EffectCategory {
+            Beneficial,
+            Harmful,
+            Neutral,
+        }
+
+        /// Represents an effect from the game
+        #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+        pub enum Effect {
+            #(#effect_variants,)*
+        }
+
+        impl Effect {
+            /// Constructs a effect from a raw item ID.
+            ///
+            /// If the given ID is invalid, `None` is returned.
+            pub const fn from_raw(id: u16) -> Option<Self> {
+                match id {
+                    #effect_from_raw_id_arms
+                    _ => None
+                }
+            }
+
+            /// Gets the raw effect ID from the effect
+            pub const fn to_raw(self) -> u16 {
+                match self {
+                    #effect_to_raw_id_arms
+                }
+            }
+
+            /// Construct a effect from its snake_case name.
+            ///
+            /// Returns `None` if the name is invalid.
+            pub fn from_ident(id: Ident<&str>) -> Option<Self> {
+                match id.as_str() {
+                    #effect_from_ident_arms
+                    _ => None
+                }
+            }
+
+            /// Gets the identifier of this effect.
+            pub const fn to_ident(self) -> Ident<&'static str> {
+                match self {
+                    #effect_to_ident_arms
+                }
+            }
+
+            /// Gets the translation key of this effect.
+            pub const fn translation_key(self) -> &'static str {
```

Comment:
`self` consumes the object `self` when this function is called. It should be `&self`. Same for `category`, `color`, etc.

Reply:
GitHub Copilot momento :skull: